tools/dynamic_framework_slicer/dynamic_framework_slicer.sh

```diff
@@ -0,0 +1,80 @@
+#!/bin/bash 
+set -eu
+
+declare -a BINARIES
+IN=""
+OUT=""
+while [[ $# -gt 0 ]]; do
+  arg="$1"
+  case $arg in
+      --in)
+      readonly IN="$2"
+      shift
+      ;;
+      --out)
+      readonly OUT="$2"
+      shift
+      ;;
+      *)
+      BINARIES+=("$1")
+      ;;
+  esac
+  shift
+done
+
+if [[ "$IN" == "" || "$OUT" == "" || "$BINARIES" == "" ]]; then
+    echo "Usage: $0 --in path/to/infile --out path/to/outfile binaries..."
+    exit 1
+fi
+
+if [[ ! -f "$IN" ]]; then
+    echo "Expected regular file with --in"
+    exit 1
+fi
+
+# Strip out any unnecessary slices from embedded dynamic frameworks to save space
+
+# Gather all binary slices
+declare -a all_bin_slices
+for bin in "${BINARIES[@]}"; do
+    # Lets figure out what slices the app has
+    bin_slices=$(xcrun lipo -info "$bin" | cut -d: -f3)
+    
+    # Add the app slices to the overall list of required slices
+    for slice in $bin_slices ; do
+        # store the slice if we don't already have it
+        if [[ "${#all_bin_slices[@]}" -eq 0 || $( echo "${all_bin_slices[@]}" | grep -q "$slice") ]] ; then
+            all_bin_slices+=($slice)
+        fi
+    done
+done
+
+# Gather the slices in the framework
+declare -a framework_slices
+for slice in $(xcrun lipo -info "$IN" | cut -d: -f3); do
+    framework_slices+=($slice)
+done
+
+# Figure out what, if anything, we should strip
+declare -a slices_needed
+just_copy=1
+for slice in "${framework_slices[@]}"; do
+    if echo "${all_bin_slices[@]}" | grep -q "$slice" ; then
+        slices_needed+=($slice)
```

Comment:
Will this always end up being `slices_needed` == `all_bin_slices`? I'm wondering if `slices_needed` needs to be computed.

Reply:
Check the resolved comments (github ui isn't great for past comments), the framework could have different arches the the apps (including the apps have some arch the framework doesn't have), so the assembly is required.

Reply:
How would the app slice successfully link if the frameworks were missing those same arch slices?

Reply:
It seems like by this point we know the frameworks have no less slices, but potentially more slices.

Reply:
The framework isn't needed for those archs.  My example - the app has conditional code for arm64 to support some AR/VR feature; the armv7 binary won't have any calls to the things vended from the framework, so the framework isn't needed.  The app ends up with armv7 and arm64, but the framework only has to have arm64.

Reply:
Yeah.

Our use case is stripping simulator slices, because App Store requires this. This seems like the common case, but not the only case 😕